### ISSUE TITLE
Using AssertionConsumerServiceURL in a signed AuthRequest

### ISSUE THREAD

**Issue by stevenao** [simplesamlphp/simplesamlphp#355]
PingIdentity is allowing that. It is a sizable provider. I am assuming the SAML2 spec is allowing it. Currently, SimpleSamlPHP is not allowing it. All ACS's must match the ones in the registered sp metadata. 

code: https://github.com/simplesamlphp/simplesamlphp/blob/b5648358afea9c52afb94b631cbdf4194ec18763/modules/saml/lib/IdP/SAML2.php#L150

PingIdentity Doc: https://ping.force.com/Support/PingFederate/Administration/Specifying-Assertion-Consumer-Service-in-AuthnRequest

**Comment by jaimeperez** [comment]
Hi @stevenao!

What is it exactly that SimpleSAMLphp does not allow? If you specify an `AssertionConsumerServiceURL` in the SAML `AuthnRequest`, SimpleSAMLphp will select that endpoint to send the response to it, provided that the URL matches with one endpoint registered in the metadata for the SP, of course.

I don't understand if what you are proposing is to allow sending responses to endpoints that are **not** registered in the metadata. If that's the case, the SAML standard explicitly forbids this in the SAML 2.0 profiles document:

>514  **4.1.4.1 \<AuthnRequest\> Usage**
>...
>532  Note that if the <AuthnRequest> is not authenticated and/or integrity protected, the information in it
533  MUST NOT be trusted except as advisory. **Whether the request is signed or not, the identity provider**
534  **MUST ensure that any `<AssertionConsumerServiceURL>` or**
535  **`<AssertionConsumerServiceIndex>` elements in the request are verified as belonging to the service**
536  **provider** to whom the response will be sent. Failure to do so can result in a man-in-the-middle attack.

Therefore, SimpleSAMLphp will not allow such a behaviour, specially when such behaviour would have negative security implications. If PingIdentity is doing it, they are breaking the standard and putting their users at risk.

**Comment by thijskh** [comment]
Ping is not alone. OpenConext accepts any ACSlocation iff the AuthnRequest is signed with a valid signature for that SP.

I do get the MITM for unsigned authnrequests, but fir signed, what's the threat scenario? Afterall, the SP can always redirect the user to any URL. I wonder why this should be forbidden.

**Comment by trscavo** [comment]
> Ping is not alone. OpenConext accepts any ACSlocation iff the AuthnRequest is signed with a valid signature for that SP.

FWIW, the Shibboleth IdP works that way as well.

**Comment by stevenao** [comment]
If the spec is not allowing SP to specify custom ACS even in a signed AuthN, and given there doesn't seem a security concern, I think it comes down to what makes most sense for users. There are quit a bit of providers implementing that way. A option to allow such behavior should be considered.

**Comment by jaimeperez** [comment]
Hi guys! Thanks a lot for the feedback!

However, I'm afraid I have to disagree with you on this, for several reasons:

1. The fact that others are doing it is not an argument, and it doesn't make it right. You could use that to support an argument, but not as **the** argument. An example of this: we've all seen recently that many people around the world hide their assets offshore in Panama. The fact that they are doing it (some of them being well-known personalities, politicians, etc), neither makes it right to do it nor legalizes money laundering :wink: 

2. It's not true that there's not a security concern. As a matter of fact, there is, even when using signed requests. The SP can always redirect the user to any URL, that's true. But we are not talking about the SP here, but the IdP. If the SP decides to redirect the user to a third party URL after authentication that's its responsibility and totally fine. In the end, the SP controls the redirection and no personal data should necessarily be involved. However, if we do that in the middle of a SAML transaction, our IdP could redirect to a third party URL **during** authentication, sending personal data to an untrusted party who we neither know anything about nor have any kind of control over it.

   You could argue of course that this can't happen if the request is signed. However, the private key of the SP could have been leaked, even inadvertently, and requests crafted by an attacker could be sent to the IdP. Those requests would include a different endpoint for the `AssertionConsumerService`, and if the IdP trusts the signature blindly, that endpoint would get an assertion from the IdP containing the user's data. Note that the SP is not involved at all in the entire attack scenario, and the IdP has no means to verify *where* it is getting the request from (other than the signature, that should no longer be trusted because the private key was leaked).

   You may also argue that this is unlikely to happen, or that if this happens, attackers might have full access to the SP, where they could also compromise the TLS key to snoop on the communications, or they could see the logs or an internal database where the personal user data is stored. None of this is necessarily true, and I'll use SimpleSAMLphp as an example. People tends to follow tutorials to install the software. Those tutorials are usually aiming at *making it work*, instead of *doing it right*. Therefore, many of them forget about **securing the installation**. Other people follow the documentation, but they ignore the part where they need to set up an alias to the `www/` directory to keep the rest of the SimpleSAMLphp installation out of the web server. Because *what's the problem?*, or *nothing's gonna happen*, or *nobody will discover this is accessible*, or *I don't understand what's the risk here*, or *I don't know how to create an alias*, or *I don't even have access to the web server's configuration*, or *name your phony reason here*. This means every day, for one reason or another, I see SimpleSAMLphp installations with their entire root directory accessible through the web. The `cert/` directory included. Sometimes, even directory listing is allowed. Even if it doesn't, it's usually not that hard to guess the name of a file holding a private key. Best-case scenario, you get the private key with only one HTTP GET request. No access to the machine, no knowledge at all, no other data leaked. But the private key is in the wild and you can now sign SAML requests with it, impersonating the SP. Note, however, that the attacker has no way to modify the SP's metadata to fit his/her purposes.

   So no, it's not true there are no security concerns here.

3. Metadata serves as a trust anchor between SAML parties. If you are exchanging metadata offline, you are setting the trust on what the metadata says due to the fact that you are receiving it from a person you trust to be its owner. If you periodically fetch metadata, you are supposed to establish the trust not only by verifying the signature of the metadata (if any) but also by verifying the protected channel where you got it from. Either way, **how can you trust something** (an unknown ACS endpoint in a SAML request) **that your main source of trust** here, the metadata, **is explicitly contradicting**?

4. Metadata is also the way you tell others **how** to communicate with you. What's the point of having your ACSs defined in there if you are not going to use them? The whole point of including those URLs in the metadata is about telling others where to send you the assertions. If you don't want them there, why not just fixing the metadata?

   I can honestly see no valid reason to do that other than laziness. You could do that in order to track the session using GET parameters, but that will be broken with many bindings (you should not use GET and POST parameters at the same time, for example) and poses security concerns. You could want to redirect the user to another URL where login was initiated, but in that case, you should do that **after** authentication. You can always save the URL and redirect there after you are done processing, there's no need to have an application URL receive a SAML assertion (besides that having an unpredictable behaviour). In general, if your ACS URL needs to change for every authentication, there's something wrong with your implementation that you should fix.

To summarize, we have something the spec is explicitly prohibiting (unless some errata fixes that, which I'm not aware of), that raises security concerns, that goes against basic trust and interoperability rules in SAML, and that serves no legitimate purpose (**up to my knowledge**, of course). So unless someone can give a **very** good reason why this **should** be done, I'm sorry but I don't think this is a reasonable feature to ask for.

**Comment by stevenao** [comment]
Thanks for the explanation.
